airbyte-workers/src/main/java/io/airbyte/workers/internal/StateMetricsTracker.java

```diff
@@ -0,0 +1,241 @@
+    }
+    updateMaxAndMeanSeconds(startingTime, timeCommitted);
+  }
+
+  void addStateMessageToStreamToStateHashTimestampTracker(final AirbyteStateMessage stateMessage,
+                                                          final int stateHash,
+                                                          final Long epochTimeEmitted) {
+
+    final StreamDescriptor streamDescriptor = stateMessage.getStream().getStreamDescriptor();
+    final String streamNameAndNamespace = streamDescriptor.getName() + streamDescriptor.getNamespace();
+    final byte[] stateHashAndTimestamp = populateStateTimestampByteArray(stateHash, epochTimeEmitted);
+
+    if (streamStateHashesAndTimestamps.get(streamNameAndNamespace) == null) {
+      final List stateHashesAndTimestamps = new ArrayList<>();
+      stateHashesAndTimestamps.add(stateHashAndTimestamp);
+      streamStateHashesAndTimestamps.put(streamNameAndNamespace, stateHashesAndTimestamps);
+    } else {
+      final List<byte[]> streamDescriptorValue = streamStateHashesAndTimestamps.get(streamNameAndNamespace);
+      streamDescriptorValue.add(stateHashAndTimestamp);
+    }
+  }
+
+  void updateMaxAndMeanSeconds(final LocalDateTime startingTime, final LocalDateTime timeCommitted) {
+    final Long secondsUntilCommit = calculateSecondsBetweenStateEmittedAndCommitted(startingTime, timeCommitted);
+    if (maxSecondsBetweenStateMessageEmittedandCommitted < secondsUntilCommit) {
+      maxSecondsBetweenStateMessageEmittedandCommitted = secondsUntilCommit;
+    }
+
+    if (totalDestinationEmittedStateMessages.get() == 1) {
+      meanSecondsBetweenStateMessageEmittedandCommitted = secondsUntilCommit;
+    } else {
+      final Long newMeanSeconds =
+          calculateMean(meanSecondsBetweenStateMessageEmittedandCommitted, totalDestinationEmittedStateMessages.get(), secondsUntilCommit);
+      meanSecondsBetweenStateMessageEmittedandCommitted = newMeanSeconds;
+    }
+  }
+
+  private LocalDateTime findStartingTimeStampAndRemoveOlderEntries(final List<byte[]> stateList, final int stateHash) {
+    // iterate through each [state_hash, timestamp] in the list
+    // update the first timestamp to equal min_timestamp
+    // and remove all items from the list as we iterate through
+    // break once we reach the state hash equal to the input(destination) state hash
+    Long minTime = null;
+    final Iterator<byte[]> iterator = stateList.iterator();
+    while (iterator.hasNext()) {
+      final byte[] stateMessageTime = iterator.next();
+      final ByteBuffer current = ByteBuffer.wrap(stateMessageTime);
+      remainingCapacity += current.capacity();
+      final int currentStateHash = current.getInt();
+      final Long epochTime = current.getLong();
+      if (minTime == null) {
+        minTime = epochTime;
+      }
+      iterator.remove();
+
+      if (stateHash == currentStateHash) {
+        break;
+      }
+    }
+    return LocalDateTime.ofEpochSecond(minTime, 0, ZoneOffset.UTC);
```

Comment:
Should we be concerned about the empty list here?

Reply:
hmmm let me think this through. we should protect against it definitely, but need to think through the behavior we want

Reply:
I updated this and added a new kind of error message for when we aren't able to match a destination message to a state message. this error is thrown when either there are no state messages in the list, or when we go through the whole list and can't find the corresponding state message.